game-core/src/main/java/games/strategy/triplea/delegate/battle/MustFightBattle.java

```diff
@@ -268,7 +268,7 @@ void addDependentUnits(final Map<Unit, Collection<Unit>> dependencies) {
   @Override
   public List<Unit> getRemainingAttackingUnits() {
     final List<Unit> remaining = new ArrayList<>(attackingUnitsRetreated);
-    final Collection<Unit> unitsLeftInTerritory = battleSite.getUnits();
+    final Collection<Unit> unitsLeftInTerritory = new ArrayList<>(battleSite.getUnits());
```

Comment:
side-note, not to update for this PR: I think one problem we are running into is a lack of encapsulation. We would not have to worry as much about set mutability if `battleSite` perhaps just handled this, eg: `battleSite.getRemainingUnits`, instead it seems we have relatively procedural code and `battleSite` is more a struct, so we run into conflicts with not trying to mutate the state of an object that is fundamentally mutable state.